### PR TITLE
[Dashboard | SDK] Feature: Ref values for address params in publish

### PR DESCRIPTION
**Pull request by kumaryash90**
PROT-932
PROT-864

## Problem solved

https://linear.app/thirdweb/project/[contract-tooling]-deployment-publish-flow-revamp-0194e04e2a84/overview

Handle contract refs in publish form -- Resolve / deploy all linked contracts in publish metadata, and pass the addresses as constructor args for corresponding params. 

It recursively deploys all referenced contracts through the ref chain, before deploying the main contract.

---

## PR-Codex overview
This PR focuses on enhancing the deployment and management of smart contracts by introducing dynamic parameters and improving UI components for contract publishing, including better handling of reference contracts and constructor parameters.

### Detailed summary
- Changed error handling in `deploy-with-abi.ts` to return address instead of throwing an error.
- Updated UI layout in `string-input.tsx` for improved responsiveness.
- Added `dynamicValue` support in various components for constructor parameters.
- Enhanced `FormFieldSetup` to accept optional `className`.
- Introduced new `DynamicParams` type for flexible contract deployment.
- Implemented dynamic input handling for reference contracts in multiple components.
- Added handling for `implConstructorParams` in contract deployment logic.
- Improved component structure for better readability and maintainability.

> The following files were skipped due to too many changes: `apps/dashboard/src/components/contract-components/contract-publish-form/contract-params-fieldset.tsx`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by gregfromstl** [comment]
@kumaryash90 could you link a linear ticket in the description? Need context to review this

**Comment by kumaryash90** [comment]
> @kumaryash90 could you link a linear ticket in the description? Need context to review this

https://linear.app/thirdweb/project/[contract-tooling]-deployment-publish-flow-revamp-0194e04e2a84/overview

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 1]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
   salt?: string;
 };
 
+interface DynamicParams {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:132]: use type everywhere

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 2]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+}
+
+type ProcessRefDeploymentsOptions = {
+  client: ThirdwebClient;
+  chain: Chain;
+  account: Account;
+  paramValue: string | ImplementationConstructorParam;
+};
+
+async function processRefDeployments(
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:161]: add unit test

Comment (joaquim-verges) [reply packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:161]: move to its own file too pls

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 3]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+            contracts[0]?.salt && contracts[0]?.salt.length > 0
+              ? contracts[0]?.salt
+              : undefined;
+
+          const addr = await deployPublishedContract({
+            client,
+            chain,
+            account,
+            contractId: contracts[0]?.contractId as string,
+            publisher: contracts[0]?.publisherAddress,
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:191]: throw clear err if none of those are defined

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 4]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+                return { name: "", type: t };
+              }),
+              values,
+            );
+          }
+        }
+
+        if (dynamicValue.type === "bytes[]") {
+          const bytesArray = [];
+          const decodedBytesArray = dynamicValue.decodedBytes;
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:257]: paramsToEncode ?

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 5]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+              const types = [];
+              const values = [];
+              for (const v of decodedBytes) {
+                types.push(v.type);
+
+                if (v.defaultValue) {
+                  values.push(v.defaultValue);
+                }
+
+                if (v.dynamicValue) {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:272]: else ?

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 6]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+                      paramValue: v,
+                    }),
+                  );
+                }
+              }
+
+              bytesArray.push(
+                encodeAbiParameters(
+                  types.map((t) => {
+                    return { name: "", type: t };
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:287]: remove name

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 7]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+
+          return JSON.stringify(bytesArray);
+        }
+      }
+    }
+    // biome-ignore lint/suspicious/noExplicitAny: catch multiple errors
+  } catch (e: any) {
+    const err = "error" in e ? e.error?.message : e?.message;
+
+    if (err.toString().includes("Contract already deployed")) {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:303]: remove

File: packages/thirdweb/src/extensions/prebuilts/deploy-published.ts [hunk 8]
```diff
@@ -128,6 +129,187 @@ export type DeployContractfromDeployMetadataOptions = {
...
+                  types.map((t) => {
+                    return { name: "", type: t };
+                  }),
+                  values,
+                ),
+              );
+            }
+          }
+
+          return JSON.stringify(bytesArray);
```
Comment (joaquim-verges) [inline packages/thirdweb/src/extensions/prebuilts/deploy-published.ts:295]: dont think you need this?

File: packages/thirdweb/src/utils/any-evm/deploy-metadata.ts [hunk 9]
```diff
@@ -159,6 +159,27 @@ type ParsedCompilerMetadata = {
   zk_version?: string;
 };
 
+type DynamicParams = {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/utils/any-evm/deploy-metadata.ts:162]: export this and re-use?

**Review by joaquim-verges — APPROVED** [review]

**Review by joaquim-verges — CHANGES_REQUESTED** [review]
shared changes / issues in DMs

File: apps/dashboard/src/components/contract-components/contract-deploy-form/custom-contract.tsx [hunk 10]
```diff
@@ -412,11 +437,12 @@ export const CustomContractForm: React.FC<CustomContractFormProps> = ({
           params: {
             name: params.contractMetadata?.name || "",
             contractURI: _contractURI,
-            defaultAdmin: params.deployParams._defaultAdmin,
+            defaultAdmin: params.deployParams._defaultAdmin as string,
```
Comment (joaquim-verges) [inline apps/dashboard/src/components/contract-components/contract-deploy-form/custom-contract.tsx:440]: always dangerous to cast like this, produces unknown behavior downstream.

would do default value or throw if you always expect it to be filled. better to fail early

**Review by joaquim-verges — APPROVED** [review]

File: apps/dashboard/src/components/contract-components/contract-deploy-form/custom-contract.tsx (hunk 10 above)
Comment (kumaryash90) [reply apps/dashboard/src/components/contract-components/contract-deploy-form/custom-contract.tsx:440]: Updated.
Only using this for special cases like marketplace or for params that can't have dynamic value.

**Comment by kumaryash90** [comment]
### Merge activity

* **Dec 11, 1:05 PM EST**: The merge label 'merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=thirdweb-dev&repo=js) once it meets the requirements.

**Pull request merged**